src/lib/aggregator.ts

```diff
@@ -1,58 +1,144 @@
 const {InvalidAggregationParams} = ERRORS;
-const {INVALID_AGGREGATION_METHOD, METRIC_MISSING} = STRINGS;
+const {INVALID_AGGREGATION_METHOD} = STRINGS;
 
 /**
  * Validates metrics array before applying aggregator.
  * If aggregation method is `none`, then throws error.
  */
-const checkIfMetricsAreValid = (metrics: string[], parameters: Parameters) => {
-  metrics.forEach(metric => {
-    const method = getAggregationMethod(metric, parameters);
-
+const checkIfMetricsAreValid = (context: ManifestCommon) => {
+  context.aggregation?.metrics.forEach(metric => {
+    const method = getAggregationMethod(metric, PARAMETERS);
     if (method === 'none') {
       throw new InvalidAggregationParams(INVALID_AGGREGATION_METHOD(method));
     }
   });
 };
 
-/**
- * Aggregates child node level metrics. Validates if metric aggregation type is `none`, then rejects with error.
- * Otherwise iterates over inputs by aggregating per given `metrics`.
- */
-export const aggregate = (
-  inputs: PluginParams[],
-  metrics: string[],
-  parameters: Parameters
-) => {
-  checkIfMetricsAreValid(metrics, parameters);
-
-  return inputs.reduce((acc, input, index) => {
-    for (const metric of metrics) {
-      if (!(metric in input)) {
-        throw new InvalidAggregationParams(METRIC_MISSING(metric, index));
+/**Returns Boolean indicating whether a node has children */
+function hasChildren(node: Node) {
+  if (node['children']) {
+    return true;
+  }
+  return false;
+}
+
+/** Returns children of given node */
+function getChildren(node: Node): Node[] {
+  const children: Node[] = [];
+  Object.keys(node.children).forEach(child => {
+    children.push(node.children[child]);
+  });
+  return children;
+}
+
+/** element-wise addition of two arrays */
+function aggregateArrays(a: Array<any>, b: Array<any>, method: string) {
+  if (method === 'sum') {
+    return a.map((element, idx) => {
+      return element + b[idx];
+    });
+  } else if (method === 'avg') {
+    return a.map((element, idx) => {
+      return (element + b[idx]) / 2;
+    });
+  }
+  return a;
+}
+
+/** aggregates multiple time series into one single time series*/
+export const aggregate = (tree: Node, context: ManifestCommon) => {
+  checkIfMetricsAreValid(context);
+  // check if root has children
+  if (hasChildren(tree)) {
+    // iterate through child nodes
+    getChildren(tree).forEach(node => {
+      // if the child has children, start recursive call to aggregate()
+      if (hasChildren(node)) {
+        aggregate(node, context);
       }
+      context.aggregation?.metrics.forEach(metric => {
+        let treeTimeSeries = tree.inputs?.map(input => input[`${metric}`]);
 
-      const accessKey = `aggregated-${metric}`;
-      acc[accessKey] = acc[accessKey] ?? 0;
-      acc[accessKey] += parseFloat(input[metric]);
+        if (
+          treeTimeSeries === undefined ||
+          treeTimeSeries?.includes(undefined)
+        ) {
+          treeTimeSeries = new Array(3).fill(0);
```

Comment:
Ahh I see, so if the parent node doesn't have inputs we just create zeros (but specifically an array of 3?) and this would make the aggregate function work logically.

Reply:
yes, exactly the 3 is a typo - it should be grabbbing the length of the node input array